py/lexer: change SyntaxError line number on missing ':'

MathieuNumworks opened the pull request:
This PR addresses https://github.com/micropython/micropython/issues/12792.

Previously, multiple newlines are merged and emmit one new line token, with the position the first character which is not a character/tab/newline/comment. 

Multiple newlines are still merged, but the position of emitted new line tokens is set to the end of the preceding line, so that a syntax error due to the lack of a token (like ':') reports to the right line.

MathieuNumworks committed:
py/lexer: Change token poisition for new lines.

Set the position of new line tokens as the end of the preceding line
instead of the beginning of the next line. This is done by first moving
the pointer to the end of the current line to skip any whitespace, record
the position for the token, then finaly skip any other line and
whitespace.
The previous behavior was to skip every new line and whitespace,
including the indent of the next line, before recording the token
position.

Signed-off-by: Mathieu Serandour <mathieu.serandour@numworks.fr>

jimmo commented:
Thanks @MathieuNumworks 

I'm not sure why the CI didn't run automatically, but I've run it now and the result is +16 bytes on stm32 (+24 on bare-arm). It's quite a good improvement for usability though!

The test failures are because there are some tests that run on coverage/unix that exact match the parser output, they should be easy to fix.

MathieuNumworks committed:
tests/cmdline/cmd_parsetree.py.exp: Fix cmd_parsetree test.

The previous commit changed the position of the new line tokens, so the
reference output for the cmd_parsetree test needed an update.

dpgeorge commented on py/lexer.c:
```diff
@@ -862,9 +872,10 @@ mp_lexer_t *mp_lexer_new(qstr src_name, mp_reader_t reader) {
+    // Check that the first token is in the first column unless it is a
+    // newline. Otherwise we convert the token kind to INDENT so that
+    // the parser gives a syntax error.
+    if (lex->tok_column != 1 && lex->tok_kind != MP_TOKEN_NEWLINE) {
```
What is this change for, is it needed?  Is there a test case that shows why it's needed?

dpgeorge replied:
Oh, I see this is needed to pass the tests, in cases where there is a comment at the start of the file.

dpgeorge commented on py/lexer.c:
```diff
@@ -531,10 +531,12 @@ STATIC bool skip_whitespace(mp_lexer_t *lex, bool stop_at_newline) {
+            if (lex->nested_bracket_level == 0) {
+                had_physical_newline = true;
+                if (stop_at_newline) {
+                    break;
```
I think it's possible to just return true here, and false at the end of the function, and not need the `had_physical_newline` variable at all.

Then this if statement doesn't need to change, it can stay as a single expression with the &&.

dpgeorge commented on py/lexer.c:
```diff
@@ -577,12 +579,20 @@ void mp_lexer_to_next(mp_lexer_t *lex) {
     lex->tok_column = lex->column;
 
+    if (had_physical_newline) {
+        next_char(lex);
```
I don't think this line is needed, the char is anyway consumed in the following call to `skip_whitespace()`.

MathieuNumworks committed:
py/lexer: Factorize conditions.

Both lex->emit_dent and lex->nested_bracket_level equal 0 if had_physical_newline == true.
The conditions can therefore be factorized.

dpgeorge commented:
I see you found a way to simplify the code a little further.  Looks really good now!

dpgeorge commented:
Rebased, squashed and merged in c85db05244ef6185fbb3c218c508ddd179830942
